### PR TITLE
display list of installations

### PR DESCRIPTION
**Pull request by SimonLab**
see #52

This PR display the list of organisations where dwylbot is installed on.

### CODE REVIEW

File: coveralls.json [hunk 1]
```diff
@@ -9,6 +9,7 @@
     "test/support/channel_case.ex",
     "test/support/conn_case.ex",
     "web/web.ex",
-    "web/views/error_helpers.ex"
+    "web/views/error_helpers.ex",
+    "web/controllers/github_api/http_client.ex"
```
Comment (samhstn) [inline coveralls.json:13]: I would probably prefer to have tests for this file

File: test/controllers/installation_controller_test.exs [hunk 2]
```diff
@@ -19,7 +19,10 @@ defmodule Dwylbot.InstallationControllerTest do
   @tag login_as: "Simon"
   test "GET /installation - list dwylbot installation", %{conn: conn} do
     conn = get conn, "/installation"
-    assert html_response(conn, 200) =~ "List of dwylbot installations"
+    response = assert html_response(conn, 200)
+    assert response =~ "List of dwylbot installations"
+    assert response =~ "dwyl"
+    assert response =~ "FocusHub"
```
Comment (shouston3) [inline test/controllers/installation_controller_test.exs]:
```elixir
for string <- ["list of dwylbot installations", "dwyl", "FocusHub"] do
  assert html_response(conn, 200) =~ string
end
```

File: web/controllers/github_api/http_client.ex [hunk 3]
```diff
@@ -0,0 +1,19 @@
+defmodule Dwylbot.GithubAPI.HTTPClient do
+  @moduledoc """
+  wrapper functions for the API Github App API
+  """
+  alias Poison.Parser, as: PP
```
Comment (samhstn) [inline web/controllers/github_api/http_client.ex:5]: Nice 👍

File: web/controllers/github_api/http_client.ex [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  defp header(token) do
+    ["Authorization": "token #{token}", "Accept": @media_type]
+  end
+
+  def get_installations(token) do
+    # create post request with header to get the list of installation of a user.
+    {:ok, res} = HTTPoison.get("https://api.github.com/user/installations", header(token), [])
+    {:ok, data} = PP.parse(res.body)
+    data["installations"]
+  end
```
Comment (samhstn) [inline web/controllers/github_api/http_client.ex]: Looks good, but I would prefer to pipe this:
```elixir
"https://api.github.com/user/installations"
|> HTTPoison.get!(header(token), [])
|> Map.fetch!(:body)
|> PP.parse!
|> Map.get("installations")
```

File: web/controllers/github_api/http_client.ex [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  """
+  alias Poison.Parser, as: PP
+
+  @media_type "application/vnd.github.machine-man-preview+json"
+
+  defp header(token) do
+    ["Authorization": "token #{token}", "Accept": @media_type]
+  end
+
+  def get_installations(token) do
```
Comment (samhstn) [inline web/controllers/github_api/http_client.ex:13]: I might add a `doc` test to this function

**Review by samhstn — COMMENTED** [review]
All Looks good, just some syntax comments

File: coveralls.json (hunk 1 above)
Comment (SimonLab) [reply coveralls.json:13]: I'm using a mock to test the Github API see https://github.com/dwyl/dwylbot/pull/54/files#diff-d9a1678446e1699ff1ed587b08c02c9f . So the mock replace the test of this file.

Comment (shouston3) [inline coveralls.json]: Missed that, nice

File: web/controllers/installation_controller.ex [hunk 6]
```diff
@@ -2,8 +2,11 @@ defmodule Dwylbot.InstallationController do
   use Dwylbot.Web, :controller
 
   plug :authenticate_user when action in [:index]
+  @github_api Application.get_env(:dwylbot, :github_api)
 
   def index(conn, _params) do
-    render conn, "index.html"
+    token = conn.assigns.current_user.token
```
Comment (samhstn) [inline web/controllers/installation_controller.ex:8]: Will `conn.assigns.current_user.token` always be defined?

File: web/controllers/github_api/http_client.ex (hunk 5 above)
Comment (samhstn) [reply web/controllers/github_api/http_client.ex:13]: Maybe not in light of https://github.com/dwyl/dwylbot/pull/54/files#r120649761

**Review by samhstn — APPROVED** [review]

File: web/controllers/installation_controller.ex (hunk 6 above)
Comment (SimonLab) [reply web/controllers/installation_controller.ex:8]: yes, the authentication plug on top of the file ([here](https://github.com/dwyl/dwylbot/pull/54/files#diff-ba9b85e7522d1e17da4c06e6cb5704c9L4)) makes sure that the index action is only accessible if the user is loggedin ie if conn.assigns.current_user is defined

**Comment by SimonLab** [comment]
Thanks for your review @shouston3 :+1: I've added the syntax changes

**Review by Danwhy — APPROVED** [review]
Looks good

**Review by nelsonic — APPROVED** [review]
@SimonLab looks great! 👍 
@Danwhy & @shouston3 thanks for peer-reviewing. 🎉

**Pull request merged**